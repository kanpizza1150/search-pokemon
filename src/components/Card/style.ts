import styled from 'styled-components'
import { setColor, setFlex } from '../../utils/globalStyles'

export const CardWrapper = styled.div`
  border-radius: 3rem;
  overflow: hidden;
  height: 70rem;
  width: 40rem;
  background-color: ${setColor.white};
  padding: 2rem;
  ${setFlex('flex-start')};
  flex-direction: column;
  margin-bottom: 3rem;
  .name {
    text-align: center;
    font-size: 2rem;
    font-weight: bold;
    padding: 1rem;
  }
  .number {
    width: 100%;
    text-align: right;
    color: ${setColor.grey};
  }
  img {
    height: 10rem;
    width: unset;
    margin: 1rem 0;
  }
  .row {
    margin-bottom: 1rem;
    width: 100%;
  }
  .tag__wrapper {
    ${setFlex('flex-start')};
    flex-wrap: wrap;
  }
`
