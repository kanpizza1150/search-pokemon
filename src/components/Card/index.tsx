import React from 'react'
import { CardWrapper, AttractWrapper, AttractDamage } from './style'
import TypeTag from '../TypeTag'
import { IPokemon } from '../../utils/interfaces'
import {
  GET_ATTRACT_BY_NAME,
  GET_EVOLUTIONS_BY_NAME,
} from '../../graphql/queries'
import { useQuery } from '@apollo/client'
import LoadingSpinner from '../LoadingSpinner'

interface IProps {
  pokemon: IPokemon
  setSearchKey: React.Dispatch<React.SetStateAction<any>>
}

const Card: React.FC<IProps> = ({ pokemon, setSearchKey }) => {
  const [cardHeight, setCardHeight] = React.useState<number>(0)
  const attractsData = useQuery(GET_ATTRACT_BY_NAME, {
    variables: { name: pokemon.name },
  })

  const evolutionData = useQuery(GET_EVOLUTIONS_BY_NAME, {
    variables: { name: pokemon.name },
  })
  const ref = React.useRef<HTMLDivElement>(null)
  // eslint-disable-next-line
  React.useEffect(() => {
    if (ref.current) {
      setCardHeight(ref.current.clientHeight)
    }
  })

  const _renderError = (): JSX.Element => (
    <div data-testid='error-mesg'>❌Fetching Data Error...❌</div>
  )

  const _renderType = (types): JSX.Element => (
    <div className='tag__wrapper'>
      {types.map((type) => (
        <TypeTag key={type} type={type} />
      ))}
    </div>
  )

  const _renderAttract = (data): JSX.Element => (
    <div className='wrapper' key={data.id}>
      {data.name} |
      <TypeTag type={data.type} /> |{' '}
      <AttractDamage type={data.type}>Damage: {data.damage}</AttractDamage>
    </div>
  )

  const onEvoNameClicked = (name: string): void => {
    setSearchKey(name)
  }

  const _renderEvo = (data, index): JSX.Element => (
    <>
      {index > 0 ? <div className='evo__arrow'>➜</div> : ''}
      <button
        className='evo__name'
        key={data.id}
        onClick={() => onEvoNameClicked(data.name)}
        data-testid={data.name}
      >
        {data.name}
      </button>
    </>
  )

  const _handleAttractsData = (): JSX.Element => {
    const { loading, data, error } = attractsData
    let result: JSX.Element = <></>
    if (loading) {
      result = <LoadingSpinner height='20rem' />
    } else if (error) {
      result = _renderError()
    } else if (data) {
      result = (
        <AttractWrapper data-testid='attracts'>
          Fast:
          <div className='section__sub'>
            {data.pokemon.attacks.fast.map((fast) => _renderAttract(fast))}
          </div>
          <div className='section__sub'>
            Special:
            {data.pokemon.attacks.special.map((special) =>
              _renderAttract(special)
            )}
          </div>
        </AttractWrapper>
      )
    }
    return result
  }

  const _handleEvolutionsData = (): JSX.Element => {
    const { loading, data, error } = evolutionData
    let result: JSX.Element = <></>
    if (loading) {
      result = <LoadingSpinner height='20rem' />
    } else if (error) {
      result = _renderError()
    } else if (data) {
      const { evolutions } = data.pokemon
      result = evolutions ? (
        <div className='evo__wrapper'>
          {data.pokemon.evolutions.map((evo, index) => _renderEvo(evo, index))}
        </div>
      ) : (
        <>-</>
      )
    }
    return result
  }

  const _renderDataRow = (): JSX.Element[] => {
    const sections = [
      {
        title: 'Types',
        key: 'type',
        value: pokemon.types,
      },
      {
        title: 'Resistant',
        key: 'resistant',
        value: pokemon.resistant,
      },
      {
        title: 'Weaknesses',
        key: 'weaknesses',
        value: pokemon.weaknesses,
      },
    ]
    return sections.map((section) => (
      <div className='row' key={section.key} data-testid={section.key}>
        <div className='title'>{section.title}: </div>
        {_renderType(section.value)}
      </div>
    ))
  }

  const _renderDataWrapper = (): JSX.Element[] => {
    const sections = [
      {
        title: 'Attract',
        key: 'attract',
        value: _handleAttractsData(),
      },
      {
        title: 'Evolutions',
        key: 'evolutions',
        value: _handleEvolutionsData(),
      },
    ]
    return sections.map((section) => (
      <div
        className='data__wrapper att'
        key={section.key}
        data-testid={section.key}
      >
        <div className='title'>{section.title}: </div>
        {section.value}
      </div>
    ))
  }
  return (
    <CardWrapper image={pokemon.image} ref={ref} cardHeight={cardHeight}>
      <div className='name' data-testid='info'>
        {pokemon.name}
        <div className='classification'>{pokemon.classification}</div>
        <div className='info'>
          <div className='info__div'>
            <div className='info__label'>Flee Rate: </div>
            <div className='info__value'>{pokemon.fleeRate}</div>
          </div>
          <div className='info__div'>
            <div className='info__label'> HP: </div>
            <div className='info__value'>{pokemon.maxHP}</div>
          </div>
          <div className='info__div'>
            <div className='info__label'> CP: </div>
            <div className='info__value'>{pokemon.maxCP}</div>
          </div>
        </div>
      </div>
      <div className='img' data-testid='image' />
      <div className='number' data-testid='number'>
        #{pokemon.number}
      </div>
      <div className='data__wrapper'>{_renderDataRow()}</div>
      {_renderDataWrapper()}
      <div className='info footer'>
        <div className='info__div footer' data-testid='height'>
          <div className='info__label'> Heigh: </div>
          <div className='info__value'>
            {pokemon.height.minimum} ~ {pokemon.height.maximum}
          </div>
        </div>
        <div className='info__div footer' data-testid='weight'>
          <div className='info__label'> Weight: </div>
          <div className='info__value'>
            {pokemon.weight.minimum} ~ {pokemon.weight.maximum}
          </div>
        </div>
      </div>
    </CardWrapper>
  )
}
export default Card
