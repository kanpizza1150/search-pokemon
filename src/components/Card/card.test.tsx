import { fireEvent, render } from '@testing-library/react'
import Card from './index'

let mockedUseQuery = {
  loading: false,
  data: {},
  error: {},
}
jest.mock('@apollo/client', () => ({
  useQuery: () => mockedUseQuery,
  gql: jest.fn(),
  ApolloProvider: jest.fn((props) => (
    <apolloprovider>{props.children}</apolloprovider>
  )),
}))

// jest.mock('../LoadingSpinner', () =>
//   jest.fn((): => <loader data-testid='loader' />)
// )
// jest.mock('../TypeTag', () => jest.fn(() => <tag data-testid='tag' />))

let mockedTypes = ['Dragon']
let mockedResistant = ['Water', 'Electric']
let mockedWeaknesses = ['Fighting', 'Fairy']
const mockedSetSearchKey = jest.fn()
describe('Card Component', () => {
  const mockedPokemon = {
    id: 'id',
    number: '001',
    name: 'name',
    classification: 'classification',
    types: [],
    resistant: [],
    weaknesses: [],
    image: 'string',
    weight: {
      minimum: 1,
      maximum: 1,
    },
    height: {
      minimum: 1,
      maximum: 1,
    },
    fleeRate: 1,
    maxCP: 1,
    maxHP: 1,
  }

  beforeEach(() => {})

  it('Should match snapshot', () => {
    const { asFragment } = render(<Card pokemon={mockedPokemon} />)
    expect(asFragment).toMatchSnapshot()
  })

  describe('Attracts section', () => {
    it('Should render loading state', () => {
      mockedUseQuery.loading = true
      const { getAllByTestId } = render(<Card pokemon={mockedPokemon} />)
      expect(getAllByTestId('loader').length).toBe(2)
    })
    it('Should render error state', () => {
      mockedUseQuery.loading = false
      mockedUseQuery.error = true
      const { getAllByTestId } = render(<Card pokemon={mockedPokemon} />)
      expect(getAllByTestId('error-mesg').length).toBe(2)
    })
    it('Should render correct data', () => {
      mockedUseQuery.loading = false
      mockedUseQuery.error = false
      mockedUseQuery.data = {
        pokemon: {
          attacks: {
            fast: [{ id: '1', name: 'name', type: 'Grass', damage: 20 }],
            special: [{ id: '2', name: 'name', type: 'Bug', damage: 21 }],
          },
        },
      }
      const { getByTestId } = render(<Card pokemon={mockedPokemon} />)
      expect(getByTestId('attracts').textContent).toBe(
        'Fast:name |Grass | Damage: 20Special:name |Bug | Damage: 21'
      )
    })
  })
  describe('Evolution section', () => {
    it('Should render loading state', () => {
      mockedUseQuery.loading = true
      const { getAllByTestId } = render(<Card pokemon={mockedPokemon} />)
      expect(getAllByTestId('loader').length).toBe(2)
    })
    it('Should render error state', () => {
      mockedUseQuery.loading = false
      mockedUseQuery.error = true
      const { getAllByTestId } = render(<Card pokemon={mockedPokemon} />)
      expect(getAllByTestId('error-mesg').length).toBe(2)
    })
    it('Should render correct data', () => {
      mockedUseQuery.loading = false
      mockedUseQuery.error = false
      mockedUseQuery.data = {
        pokemon: {
          ...mockedUseQuery.data.pokemon,
          evolutions: [
            { id: '1', name: 'evo-1' },
            { id: '2', name: 'evo-2' },
          ],
        },
      }
      const { getByTestId } = render(<Card pokemon={mockedPokemon} />)
      expect(getByTestId('evolutions').textContent).toBe(
        'Evolutions: evo-1âžœevo-2'
      )
    })
    it('Should click on evo name button', () => {
      const { getByTestId } = render(
        <Card pokemon={mockedPokemon} setSearchKey={mockedSetSearchKey} />
      )
      fireEvent.click(getByTestId('evo-1'))
      expect(mockedSetSearchKey).toBeCalledWith('evo-1')
    })
  })

  describe('Info section', () => {
    it('Should render card correctly', () => {
      const { getByTestId } = render(<Card pokemon={mockedPokemon} />)
      expect(getByTestId('info').textContent).toBe(
        'nameclassificationFlee Rate: 1 HP: 1 CP: 1'
      )
      expect(getByTestId('number').textContent).toBe('#001')
      expect(getByTestId('height').textContent).toBe(' Heigh: 1 ~ 1')
      expect(getByTestId('weight').textContent).toBe(' Weight: 1 ~ 1')
    })
    it('Should render Resistant correctly', () => {
      mockedPokemon.resistant = mockedResistant
      const { getByTestId } = render(<Card pokemon={mockedPokemon} />)
      expect(getByTestId('tag-Water')).toBeInTheDocument()
      expect(getByTestId('tag-Electric')).toBeInTheDocument()
    })
    it('Should render Weaknesses correctly', () => {
      mockedPokemon.weaknesses = mockedWeaknesses
      const { getByTestId } = render(<Card pokemon={mockedPokemon} />)
      expect(getByTestId('tag-Fighting')).toBeInTheDocument()
      expect(getByTestId('tag-Fairy')).toBeInTheDocument()
    })
    it('Should render Type correctly', () => {
      mockedPokemon.types = mockedTypes
      const { getByTestId } = render(<Card pokemon={mockedPokemon} />)
      expect(getByTestId('tag-Dragon')).toBeInTheDocument()
    })
  })
})
