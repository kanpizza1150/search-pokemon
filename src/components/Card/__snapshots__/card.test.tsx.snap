// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component Should match snapshot 1`] = `[Function]`;
